### PR TITLE
Add fully parameterized manual contact tracing

### LINKED ISSUE #84: Support Manual Contact Tracing
Support traditional contact tracing within the simulation, where traces can occur across contacts even if neither have the app installed, bounded by the capacity of the testers.

An example of the proposed modifications:

From Hospitalization or Positive Test, start manual contact tracing. Trace the users through existing contacts. Begin quarantine process, including testing on trace/symptoms as usual. Results on positive/hospitalization proceed back through the start of this loop.

Flags:
* manual_tracing_on(_time) - bool, use any manual tracing (default false)
* manual_trace_on_hospitalization - bool, start a trace on hospitalization (default true)
* manual_trace_on_positive - bool, start a trace on positive test (default false)
* manual_trace_delay - int, number of days between trigger and trace.
* manual_traceable_fraction[] - float [0,1], per network, we control the likelihood of a successful trace, especially as home (and work to a lesser extent) networks are far more likely to have successful manual tracing due to better knowledge of the contacts. (default 1 for home, 0.8 for work/school, 0.05 for random). Another option is to break this out to account for failures to contact an index case at all (e.g. manual_trace_interview_fraction), which could vary for positive/hospital as well.
* manual_trace_exclude_app_users - bool, only trace on people without digital contact tracing app. (default false)
* manual_trace_n_workers - int, size of worker pool (alternately could be _per_100k, as per guidelines, scaled based on n_total) (default 30 per 100k, per [1,2])
* manual_trace_interviews_per_worker_day - int, number of index case interviews each worker can place per day to start tracing (default 6, per [3]). Can be set arbitrarily high to allow for manually tracing all possible cases.
* manual_trace_notifications_per_worker_day - int, number of tracing notification calls each worker can place per day (not including follow-up calls) (default 12, per [3]). Can be set arbitrarily high to allow for manually tracing all possible cases.
* manual_trace_quarantine_fraction - float [0,1], likelihood of someone quarantining based on manual tracing (optional, can use existing to start) (default = 0.8)

[1] https://www.npr.org/sections/health-shots/2020/04/28/846736937/we-asked-all-50-states-about-their-contact-tracing-capacity-heres-what-we-learne
[2] https://www.naccho.org/uploads/full-width-images/Contact-Tracing-Statement-4-16-2020.pdf
[3] https://www.gwhwi.org/estimator-613404.html

### PR DESCRIPTION
**Pull request by mattea**
Add fully parameterized manual contact tracing.

All the commits are collapsed as I had to re-apply the changes to work with the history changes of the May 29 commits.

If you'd rather the functions be separated as in the related PR, I'm happy to refactor. This is just the original way I wrote it.

Fixes #84

### CODE REVIEW

File: tests/test_interventions.py [hunk 1]
```diff
@@ -1204,6 +1289,39 @@ def test_quarantined_have_trace_token(self, test_params, time_steps_test ):
...
+        index_traced = index_traced[ index_traced[ "index_ID" ] != index_traced[ "traced_ID" ] ]
+        np.testing.assert_equal( len( index_traced ) > 0, True, "no tracing has occured") 
+        
+        df_indiv.rename( columns = { "ID":"index_ID" }, inplace = True )
+        test = pd.merge( index_traced, df_indiv, on = "index_ID", how = "left")
+        np.testing.assert_equal( len( test[ test[ "app_user" ] != 1 ] ) > 0, True, "no manual tracing started" ) 
+        
+        df_indiv.rename( columns = { "index_ID":"traced_ID" }, inplace = True )
+        test = pd.merge( index_traced, df_indiv, on = "traced_ID", how = "left")
+        np.testing.assert_equal( len( test[ test[ "app_user" ] != 1 ] ) > 0, True, "no manual traces" )
```
Comment (roberthinch) [inline tests/test_interventions.py:1326]: Given the large number of options, you need to have more features tested. Specifically you need to test at least whether:

1. Delays in manual tracing are working (currently not)
2. Trace proportion by network.

File: src/interventions.c [hunk 2]
```diff
@@ -339,6 +342,15 @@ void update_intervention_policy( model *model, int time )
 
 	if( time == params->testing_symptoms_time_off )
 		set_model_param_test_on_symptoms( model, FALSE );
+
+	if( time == params->testing_symptoms_time_off )
+		set_model_param_test_on_symptoms( model, FALSE );
+
+	if (params->manual_trace_on)
```
[the remark this answers is not in the document]
Comment (mattea) [reply src/interventions.c]: Done

File: src/interventions.c [hunk 3]
```diff
@@ -535,12 +559,22 @@ void intervention_notify_contacts(
...
 				contact = inter->individual;
-				if( contact->app_user )
+				if( contact->app_user || trace_type == MANUAL_TRACE )
 				{
 					if( inter->traceable == UNKNOWN )
-						inter->traceable = gsl_ran_bernoulli( rng, params->traceable_interaction_fraction );
+					{
+						if( contact->app_user && trace_type == DIGITAL_TRACE )
+							inter->traceable = gsl_ran_bernoulli( rng, params->traceable_interaction_fraction );
+						else if ( trace_type == MANUAL_TRACE && model->manual_trace_notification_quota > 0 )
```
[the remark this answers is not in the document]
Comment (mattea) [reply src/interventions.c]: Done

File: src/interventions.c [hunk 4]
```diff
@@ -535,12 +559,22 @@ void intervention_notify_contacts(
...
+							inter->traceable = gsl_ran_bernoulli( rng, params->manual_traceable_fraction[inter->type] );
```
[the remark this answers is not in the document]
Comment (mattea) [reply src/interventions.c]: Done.

File: src/interventions.c [hunk 5]
```diff
@@ -535,12 +559,22 @@ void intervention_notify_contacts(
...
+					}
 					if( inter->traceable )
-						intervention_on_traced( model, contact, model->time - ddx, recursion_level, index_token, risk_scores[ contact->age_group ] );
+					{
+						if (trace_type == MANUAL_TRACE)
```
[the remark this answers is not in the document]
Comment (mattea) [reply src/interventions.c]: Done

File: src/interventions.c [hunk 6]
```diff
@@ -636,7 +670,10 @@ void intervention_quarantine_household(
 			intervention_quarantine_until( model, contact, time_event, TRUE, index_token, contact_time, risk_scores[ contact->age_group ] );
 
 			if( contact_trace && ( model->params->quarantine_on_traced || model->params->test_on_traced ) )
-				intervention_notify_contacts( model, contact, NOT_RECURSIVE, index_token );
+			{
+				intervention_notify_contacts( model, contact, NOT_RECURSIVE, index_token, DIGITAL_TRACE );
+				intervention_notify_contacts( model, contact, NOT_RECURSIVE, index_token, MANUAL_TRACE );
```
[the remark this answers is not in the document]
Comment (mattea) [reply src/interventions.c]: Done

File: src/interventions.c [hunk 7]
```diff
@@ -831,7 +868,13 @@ void intervention_on_positive_result( model *model, individual *indiv )
 	 ( !index_already || !params->trace_on_symptoms || params->retrace_on_positive ) &&
 	  ( params->quarantine_on_traced || params->test_on_traced )
 	)
-		intervention_notify_contacts( model, indiv, 1, index_token );
+		intervention_notify_contacts( model, indiv, 1, index_token, DIGITAL_TRACE );
+
+	if( ( params->manual_trace_on_positive ||
+		  ( params->manual_trace_on_hospitalization && is_in_hospital( indiv ) ) ) &&
```
[the remark this answers is not in the document]
Comment (mattea) [reply src/interventions.c:969]: Correct, it's just to allow for not tracing on any positive if you only want to trace on hospitalization (some existing manual tracing scenarios do this).

File: src/interventions.c [hunk 8]
```diff
@@ -886,7 +931,8 @@ void intervention_on_traced(
 
 	if( params->quarantine_on_traced )
 	{
-		int time_event = model->time;
+		int trace_time = (trace_type == DIGITAL_TRACE) ? 0 : model->params->manual_trace_delay;
```
[the remark this answers is not in the document]
Comment (mattea) [reply src/interventions.c]: Fixed.

File: tests/test_interventions.py (hunk 1 above)
Comment (mattea) [reply tests/test_interventions.py]: Done. I'm not happy with exactly how the tests work, but they should cover the important scenarios without needing to add too many more tracking variables within the simulation. But we can add more if we want to do more precise accounting.

**Comment by roberthinch** [comment]
I changed the test_manual_trace_params to use trace tokens and then tested that manual tracing occurred at exactly the correct time for everyone. The problem with using quarantine reasons is that for each person there can be multiple reasons for quarantining so it is hard to disentangle (hence the previous test used the delay with the most number of people instead of the the delay being exactly one date).

We should probably have a test on the quotas as well, not sure how exactly to do it.

**Pull request merged**